apps/docs/pages/guides/auth/auth-hooks.mdx

```diff
@@ -0,0 +1,717 @@
+Hooks help you:
+
+- Track the origin of user signups by adding metadata
+- Improve security by adding additional checks to password and multi-factor authentication
+- Support legacy systems by integrating with identity credentials from external authentication systems
+- Add additional custom claims to your JWT
+
+You can use hooks at specific points along an Auth flow to perform custom behavior via Postgres Functions. The following hooks are available:
+
+| Hook                          | Available on Plan    |
+| ----------------------------- | -------------------- |
+| Custom Access Token           | Free, Pro            |
+| MFA Verification Attempt      | Teams and Enterprise |
+| Password Verification Attempt | Teams and Enterprise |
+
+You can connect a hook to Supabase Auth to signal to Supabase Auth that it should make use of the hook.
+
+Below is a flow chart illustrating how Supabase Auth uses an Auth Hook:
+
+<Image
+  alt="Auth Hooks Flowchart"
+  width={'1000'}
+  height={'400'}
+  src="/docs/img/guides/auth-hooks/hooks-flowchart.png"
+/>
+
+## Create a Hook
+
+### What is a Hook
+
+A hook is a [Postgres Function](https://www.postgresql.org/docs/current/sql-createfunction.html) that takes in a `jsonb` object and return a `jsonb` object. You can transform the `jsonb` input in the body of the hook via native operators such as [JSON operators](https://www.postgresql.org/docs/16/functions-json.html) or extensions such as [plv8](https://supabase.com/docs/guides/database/extensions/plv8).
+
+<Admonition type="caution">
+  The `?` JSONB operator is not supported as it conflicts with our ORM.
+</Admonition>
+
+Here is an example hook signature:
+
+```sql
+create or replace function public.custom_access_token_hook(event jsonb)
+returns jsonb
+language plpgsql
+as $$
+declare
+  -- Insert variables here
+begin
+  -- Insert logic here
+  return event;
+end;
+$$;
+```
+
+You can visit `SQL Editor > Templates` for hook templates.
+
+### Hook grants and roles
+
+You need to assign additional permissions so that Supabase Auth can access the hook as well as the tables it interacts with.
+
+The `supabase_auth_admin` role does not have permissions to the `public` schema. You need to grant the role permission to execute your hook function:
+
+```sql
+grant execute
+  on function public.custom_access_token_hook
+  to supabase_auth_admin;
+```
+
+Also revoke permissions from the `authenticated` and `anon` roles to ensure the function is not accessible by Supabase Serverless APIs.
+
+```sql
+revoke execute
+  on function public.custom_access_token_hook
+  from authenticated, anon;
+```
+
+For security, we recommend against the use the `security definer` tag. The `security definer` tag specifies that the function is to be executed with the privileges of the user that owns it. When a function is created via the Supabase dashboard with the tag, it will have the extensive permissions of the `postgres` role which could allow for undesirable actions to occur.
```

Comment:
```suggestion
For security, we recommend against the use the `security definer` tag. The `security definer` tag specifies that the function is to be executed with the privileges of the user that owns it. When a function is created via the Supabase dashboard with the tag, it will have the extensive permissions of the `postgres` role which make it easier for undesirable actions to occur.
```